native/libc_test.py fails on OS X

I think these were legit differences because we were testing out a lot of fnmatch(), and the libc on OS X behaves differently

Thread on Zulip

e.g. this one https://oilshell.zulipchat.com/#narrow/stream/121540-oil-discuss/topic/building.20dev.20env.20in.20Nix.20.2B.20macOS/near/187785631

Sorry--here's the actual output :)

```
building 'libc' extension
clang -UNDEBUG -I/nix/store/sjw2xj6d38x4135gj6j8yygrpb2f39lf-python-2.7.17/include/python2.7 -c native/libc.c -o build/temp.macosx-10.12-x86_64-2.7/native/libc.o
native/libc.c:288:13: warning: using the result of an assignment as a condition without parentheses [-Wparentheses]
  if (match = (regexec(&pat, str, outlen, pmatch, 0) == 0)) {
      ~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
native/libc.c:288:13: note: place parentheses around the assignment to silence this warning
  if (match = (regexec(&pat, str, outlen, pmatch, 0) == 0)) {
            ^
      (                                                   )
native/libc.c:288:13: note: use '==' to turn this assignment into an equality comparison
  if (match = (regexec(&pat, str, outlen, pmatch, 0) == 0)) {
            ^
            ==
1 warning generated.
clang -bundle -undefined dynamic_lookup build/temp.macosx-10.12-x86_64-2.7/native/libc.o -L/nix/store/sjw2xj6d38x4135gj6j8yygrpb2f39lf-python-2.7.17/lib -o /Users/abathur/work/oil/libc.so
libc.so: Mach-O 64-bit x86_64 bundle, flags:<NOUNDEFS|DYLDLINK|TWOLEVEL>
F....real	0.100
user	0.200
sys	0.300
.....FF
======================================================================
FAIL: testFnmatch (__main__.LibcTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "native/libc_test.py", line 79, in testFnmatch
    expected, actual, '%r %r -> got %d' % (pat, s, actual))
AssertionError: '\\' '\\' -> got 1

======================================================================
FAIL: testRegexParse (__main__.LibcTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "native/libc_test.py", line 125, in testRegexParse
    self.assertRaises(RuntimeError, libc.regex_parse, '{')
AssertionError: RuntimeError not raised

======================================================================
FAIL: testWcsWidth (__main__.LibcTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "native/libc_test.py", line 199, in testWcsWidth
    self.assertEqual(1, libc.wcswidth("▶️"))
AssertionError: 1 != 2

----------------------------------------------------------------------
Ran 12 tests in 0.002s

FAILED (failures=3)
```

--- a/native/libc_test.py
+++ b/native/libc_test.py
@@ -10,9 +10,12 @@
 libc_test.py: Tests for libc.py
 """
 import unittest
+import sys
 
 import libc  # module under test
 
+# guard some tests that fail on Darwin
+IS_DARWIN = sys.platform == 'darwin'
 
 class LibcTest(unittest.TestCase):
 
@@ -55,7 +58,7 @@ class LibcTest(unittest.TestCase):
         ('strange]one', 'strange]one', 1),
 
         # What is another error?  Invalid escape is OK?
-        ('\\', '\\', 0),  # no pattern is valid
+        None if IS_DARWIN else ('\\', '\\', 0),  # no pattern is valid
 
         ('[[:alpha:]]', 'a', 1),
         ('[^[:alpha:]]', 'a', 0),  # negate
@@ -69,11 +72,11 @@ class LibcTest(unittest.TestCase):
         # Hm [] is treated as a constant string, not an empty char class.
         # Should we change LooksLikeGlob?
         ('[]', '', 0),
-        ('[]', 'a', 0),
-        ('[]', '[]', 1),
+        None if IS_DARWIN else ('[]', 'a', 0),
+        None if IS_DARWIN else ('[]', '[]', 1),
     ]
 
-    for pat, s, expected in cases:
+    for pat, s, expected in filter(None, cases):
       actual = libc.fnmatch(pat, s)
       self.assertEqual(
           expected, actual, '%r %r -> got %d' % (pat, s, actual))
@@ -122,7 +125,7 @@ class LibcTest(unittest.TestCase):
     # Syntax errors
     self.assertRaises(RuntimeError, libc.regex_parse, r'*')
     self.assertRaises(RuntimeError, libc.regex_parse, '\\')
-    self.assertRaises(RuntimeError, libc.regex_parse, '{')
+    IS_DARWIN or self.assertRaises(RuntimeError, libc.regex_parse, '{')
 
     cases = [
         ('([a-z]+)([0-9]+)', 'foo123', ['foo123', 'foo', '123']),
@@ -133,12 +136,12 @@ class LibcTest(unittest.TestCase):
         # The match is unanchored
         (r'.c', 'abcd', ['bc']),
         # Empty matches empty
-        (r'', '', ['']),
+        None if IS_DARWIN else (r'', '', ['']),
         (r'^$', '', ['']),
         (r'^.$', '', None),
     ]
 
-    for pat, s, expected in cases:
+    for pat, s, expected in filter(None, cases):
       #print('CASE %s' % pat)
       actual = libc.regex_match(pat, s)
       self.assertEqual(expected, actual)
@@ -196,8 +199,8 @@ class LibcTest(unittest.TestCase):
       print('width % d' % width)
 
   def testWcsWidth(self):
-    self.assertEqual(1, libc.wcswidth("▶️"))
-    self.assertEqual(28, libc.wcswidth("(osh) ~/.../unchanged/oil ▶️ "))
+    IS_DARWIN or self.assertEqual(1, libc.wcswidth("▶️"))
+    IS_DARWIN or self.assertEqual(28, libc.wcswidth("(osh) ~/.../unchanged/oil ▶️ "))
     self.assertEqual(2, libc.wcswidth("→ "))
     self.assertRaises(UnicodeError, libc.wcswidth, "\xfe")
 

